### PR TITLE
[KV] Add --index-only, --jdbc-url and --full-metrics to IntegrityChecker [KVL-557]

### PR DESCRIPTION
**Pull request by fabiotudone-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by tudor-da — COMMENTED** [review]
LGTM

File: ledger/participant-state/kvutils/tools/README.md [hunk 1]
```diff
@@ -26,6 +26,9 @@ The following options are supported:
     --help
     --skip-byte-comparison  Skips the byte-for-byte comparison. Useful when comparing behavior across versions.
     --sort-write-set        Sorts the computed write set. Older exports sorted before writing. Newer versions order them intentionally.
+    --index-only            Run only the indexing step of the integrity checker (useful to benchmark the indexer).
+    --jdbc-url              External JDBC url (useful for running against PostgreSQL).
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/README.md:30]:
```suggestion
    --jdbc-url              External JDBC URL (useful for running against PostgreSQL).
```

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala [hunk 2]
```diff
@@ -33,7 +36,7 @@ object Config {
         s"You can produce a ledger export on a kvutils ledger by setting ${LedgerDataExporter.EnvironmentVariableName}=/path/to/file${System.lineSeparator}")
       help("help")
       arg[Path]("PATH")
-        .text("The path to the ledger export file.")
+        .text("The path to the ledger export file (un-archived).")
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala:39]: Do you mean uncompressed?

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala [hunk 3]
```diff
@@ -128,24 +144,37 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
...
     }.transform {
-      case Success(value) =>
-        println(s"Successfully indexed all updates.".green)
-        println()
-        Success(value)
+      case Success((startTime, _)) =>
+        Success {
+          println(s"Successfully indexed all updates.".green)
+          val durationSeconds = Duration.fromNanos(System.nanoTime() - startTime).toSeconds
+          println(
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala:174]: To improve readability please move the calculation of updates/second into a `val`.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala [hunk 4]
```diff
@@ -178,29 +207,24 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
...
-              compareWriteSets(expectedWriteSet, orderedActualWriteSet)
-            } else {
-              ()
-            }
-          } yield ()
+            ) map { _ =>
+              val actualWriteSet = queryableWriteSet.getAndClearRecordedWriteSet()
+              val orderedActualWriteSet =
+                if (config.sortWriteSet) actualWriteSet.sortBy(_._1.asReadOnlyByteBuffer())
+                else actualWriteSet
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala:223]: According to the [Scala style guide](https://docs.scala-lang.org/style/control-structures.html#curly-braces):
```suggestion
                if (config.sortWriteSet)
                  actualWriteSet.sortBy(_._1.asReadOnlyByteBuffer())
                else
                  actualWriteSet
```

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala [hunk 5]
```diff
@@ -178,29 +207,24 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
...
+              actualReadServiceFactory.appendBlock(orderedActualWriteSet)
+              if (config.performByteComparison)
+                compareWriteSets(expectedWriteSet, orderedActualWriteSet)
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala:226]: According to the [Scala style guide](https://docs.scala-lang.org/style/control-structures.html#curly-braces):
```suggestion
              if (config.performByteComparison) {
                compareWriteSets(expectedWriteSet, orderedActualWriteSet)
              }
```

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala [hunk 6]
```diff
@@ -42,6 +45,16 @@ object Config {
...
           "Sorts the computed write set. Older exports sorted before writing. Newer versions order them intentionally.")
         .action((_, c) => c.copy(sortWriteSet = true))
+      opt[Unit]("index-only")
+        .text(
+          "Run only the indexing step of the integrity checker (useful tp benchmark the indexer).")
+        .action((_, c) => c.copy(indexOnly = true))
+      opt[String]("jdbc-url")
+        .text("External JDBC url (useful for running against PostgreSQL).")
+        .action((jdbcUrl, c) => c.copy(jdbcUrl = Some(jdbcUrl)))
+      opt[Unit]("full-metrics")
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala:55]: Let's name it `print-metrics`/`report-metrics` or similar then. I.e., `full-metrics` doesn't imply that we'll see these metrics at the end.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala [hunk 7]
```diff
@@ -128,24 +144,37 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
...
+        case (feedHandle, startTime) => Future.successful(startTime).zip(feedHandle.completed())
+      }
     }.transform {
-      case Success(value) =>
-        println(s"Successfully indexed all updates.".green)
-        println()
-        Success(value)
+      case Success((startTime, _)) =>
+        Success {
+          println(s"Successfully indexed all updates.".green)
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala:172]:
```suggestion
          println("Successfully indexed all updates.".green)
```

File: ledger/participant-state/kvutils/tools/README.md (hunk 1 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/README.md:30]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala (hunk 2 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala (hunk 4 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala (hunk 6 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala (hunk 7 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala [hunk 8]
```diff
@@ -12,6 +12,9 @@ case class Config(
     exportFilePath: Path,
     performByteComparison: Boolean,
     sortWriteSet: Boolean,
+    indexOnly: Boolean = false,
+    fullMetrics: Boolean = false,
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala:16]: Please make this consistent with the command-line option itself.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala [hunk 9]
```diff
@@ -41,6 +44,16 @@ object Config {
         .text(
           "Sorts the computed write set. Older exports sorted before writing. Newer versions order them intentionally.")
         .action((_, config) => config.copy(sortWriteSet = true))
+      opt[Unit]("index-only")
+        .text(
+          "Run only the indexing step of the integrity checker (useful tp benchmark the indexer).")
+        .action((_, c) => c.copy(indexOnly = true))
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala:50]: Please don't use 1-character variable names.

**Review by miklos-da — APPROVED** [review]

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala (hunk 3 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala]: Made a small change to fix this hard-to-read multi-line here.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala (hunk 8 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala (hunk 9 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/Config.scala]: Done.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala [hunk 10]
```diff
@@ -106,19 +113,25 @@ class IntegrityChecker[LogResult](commitStrategySupport: CommitStrategySupport[L
         config,
       )
       _ <- stateUpdates.compare()
```
Comment (tudor-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/IntegrityChecker.scala:115]: This shouldn't be here

**Pull request merged**